src/main/java/hudson/plugins/mercurial/MercurialSCMSource.java

```diff
@@ -190,6 +186,42 @@ public long lastModified() {
         }
     }
 
+    @Override
+    @CheckForNull
+    protected SCMRevision retrieve(@NonNull String thingName, @NonNull TaskListener listener) throws IOException, InterruptedException {
+        MercurialInstallation inst = MercurialSCM.findInstallation(installation);
+        if (inst == null) {
+            listener.error("No configured Mercurial installation");
+            return null;
+        }
+        if (!inst.isUseCaches()) {
+            listener.error("Mercurial installation " + installation + " does not support caches");
+            return null;
+        }
+        final Node node = Jenkins.getInstance();
+        if (node == null) { // Should not happen BTW
+            listener.error("Cannot retrieve the Jenkins master node");
+            return null;
+        }
+        Launcher launcher = node.createLauncher(listener);
+        StandardUsernameCredentials credentials = getCredentials();
+        final FilePath cache = Cache.fromURL(source, credentials, inst.getMasterCacheRoot()).repositoryCache(inst, node, launcher, listener, true);
+        if (cache == null) {
+            listener.error("Could not use caches, not fetching branch heads");
+            return null;
+        }
+        final HgExe hg = new HgExe(inst, credentials, launcher, node, listener, new EnvVars());
+        try {
+            String revision = hg.popen(cache, listener, true, new ArgumentListBuilder("id", "-r", thingName, "--id", "--branch"));
```

Comment:
UI code can always feel free to trim hashes down to compact forms for display, but internally we should work with full hashes.